src/glomap/processors/reconstruction_pruning.cc

```diff
@@ -1,128 +1,234 @@
+    }
+  }
+
+  LOG(INFO) << "Clustering took " << iteration << " iterations. "
+            << "Frames are grouped into " << num_valid_clusters
+            << " clusters (size >= " << kMinClusterSize << ")";
+
+  return cluster_ids;
+}
+
+}  // namespace
+
+std::unordered_map<frame_t, int> PruneWeaklyConnectedFrames(
+    colmap::Reconstruction& reconstruction) {
+  // Step 1: Compute covisibility counts between all frame pairs.
+  // For each 3D point, increment the count for every pair of frames that sees
+  // it.
+  std::unordered_map<frame_pair_t, int> frame_covisibility_count;
+  std::unordered_set<frame_t> nodes;
   for (const auto& [track_id, track] : reconstruction.Points3D()) {
     if (track.track.Length() <= 2) continue;
 
     for (size_t i = 0; i < track.track.Length(); i++) {
       const image_t image_id1 = track.track.Element(i).image_id;
       const frame_t frame_id1 = reconstruction.Image(image_id1).FrameId();
 
-      frame_observation_count[frame_id1]++;
+      nodes.insert(frame_id1);
       for (size_t j = i + 1; j < track.track.Length(); j++) {
         const image_t image_id2 = track.track.Element(j).image_id;
         const frame_t frame_id2 = reconstruction.Image(image_id2).FrameId();
         if (frame_id1 == frame_id2) continue;
-        const image_pair_t pair_id =
+        const frame_pair_t pair_id =
             colmap::ImagePairToPairId(frame_id1, frame_id2);
-        if (pair_covisibility_count.find(pair_id) ==
-            pair_covisibility_count.end()) {
-          pair_covisibility_count[pair_id] = 1;
-        } else {
-          pair_covisibility_count[pair_id]++;
-        }
+        frame_covisibility_count[pair_id]++;
       }
     }
   }
 
-  // Establish the visibility graph
-  size_t counter = 0;
-  ViewGraph visibility_graph_frame;
-  std::vector<int> pair_count;
-  for (auto& [pair_id, count] : pair_covisibility_count) {
-    // since the relative pose is only fixed if there are more than 5 points,
-    // then require each pair to have at least 5 points
-    if (count >= 5) {
-      counter++;
-      const auto [image_id1, image_id2] = colmap::PairIdToImagePair(pair_id);
-
-      if (frame_observation_count[image_id1] < min_num_observations ||
-          frame_observation_count[image_id2] < min_num_observations)
-        continue;
-
-      ImagePair image_pair;
-      image_pair.is_valid = true;
-      image_pair.weight = count;
-      visibility_graph_frame.AddImagePair(
-          image_id1, image_id2, std::move(image_pair));
-
-      pair_count.push_back(count);
-    }
-  }
-  LOG(INFO) << "Established visibility graph with " << counter << " pairs";
+  // Step 2: Filter edges to keep only reliable connections.
+  // Require at least 5 shared points (needed for stable relative pose).
+  std::unordered_map<frame_pair_t, int> edge_weights;
+  std::vector<int> weight_values;
+  for (const auto& [pair_id, count] : frame_covisibility_count) {
+    if (count < 5) continue;
```

Comment:
Suggest extracting these magical thresholds into constexpr at the beginning or even make them part of an options struct?